### ISSUE TITLE
Abnormally high VRAM usage on start

### ISSUE THREAD

**Issue by OrderlyUnicode** [hyprwm/Hyprland#2945]
### Hyprland Version

appears on 3f7f4207a6dd8bd217c01514ea40e9e4b9b7874a , the `hyprland` package from the Arch repositories, and `hyprland-git` in the AUR

### Bug or Regression?

Bug

### Description

Hyprland is utilising ~256MB of VRAM on a cold boot with the only startup apps being `hyprpaper` and `waybar`in spite of any decorations that could possibly contribute to this, such as blur or animations, being disabled. Unlike in #809 , this does not appear to be influenced by the configuration file being reloaded or windows being closed, like in #357 .(although it did seem the issues were on in the same when I read up on those) This has lead to stability issues initially believed to be related to other software, but has since been found to be an issue with the system very quickly running out of video RAM.

In case it's at all relevant, I have an AMD CPU and Nvidia GPU. the Integrated GPU allocates 1GB by default. There is an option to allocate 2GB, but it cuts into the amount of system RAM I have available.

### How to reproduce

Simply start `hyprland` and view the VRAM usage via `nvtop`

### Crash reports, logs, images, videos

I have no crash reports to show, but here's the relevant portions of my hyprland.conf file

```
animations:enabled = no
decoration:blur:enabled = no
decoration:drop_shadow = no
misc:disable_hyprland_logo = true
```

**Comment by Haudiel** [comment]
How did you fix it?

**Comment by OrderlyUnicode** [comment]
I didn't, I just happened to have an option in the BIOS that increases the VRAM allotted to my iGPU to a point where I would run into performance issues before I ran out of VRAM, and I use that as a temporary workaround. It does have the downside that I'm only able utilise 13 of the 16GB of RAM I have, which will probably become an issue in and of itself eventually.

**Comment by vaxerski** [comment]
250MB of VRAM doesn't sound like too much...?

**Comment by OrderlyUnicode** [comment]
> 250MB of VRAM doesn't sound like too much...?

It's more than double what sway uses for me

And I would agree with you if I were running this from a desktop where I could ignore the integrated GPU alltogether. If I didn't have to worry about my battery lasting for more than two hours, I would just have it use the Nvidia card with 3-6 times the VRAM, depending on a BIOS setting I had to use as a workaround.

**Comment by TristenSeth** [comment]
I have a similar issue right now on my Arch desktop. I have a ryzen 7600 and RTX 3070, but with literally just one terminal and this single firefox tab open, Hyprland is using 2.2 GB (!) of VRAM according to nvtop, and an additional 2.1GB of system RAM. What is going on here?

**Comment by vaxerski** [comment]
no clue, for me it's 250MB on both.

**Comment by TristenSeth** [comment]
Any pointers on debugging? I've read the wiki, but I'm a bit confused how i can actually debug outside of just running a nested instance or just straight up run the debug build and read logs. I was wondering how to setup GDB or Valgrind to attach to the debug build, since those are the tools i have the most experience with. I would love to dig in and see what i can find and submit a PR if its something worthwhile

**Comment by TristenSeth** [comment]
As a follow up, a clean session starts with 585MB of VRAM use with another ~528MB of system RAM. Not insane, but still a lot higher than expected i think

**Comment by TristenSeth** [comment]
Okay, i figured out how i can step through with GDB. It seems the bug is not present on the debug build, at least for the git HEAD. Perhaps it is only present on the Arch repo package. Right now, the debug build is using 75MB of VRAM with 2 windows open, and only about 62MB of system RAM. I am going to try to just build this version from source and use that for a while. I will update this with my findings.

**Comment by TristenSeth** [comment]
Okay, seems even the debug build is having this issue when it is not running as a nested instance. Looking through the logs, this seems to be the common nonoccurence when closing a window:

`[ERR] onWindowRemovedTiling node null?
[LOG] Cleanup: destroyed a window
[LOG] [wl res 55f233ef8410] destroyed (wl_resource_destroy not sent)
[LOG] [wl res 55f233f2a240] destroyed (wl_resource_destroy not sent)
[LOG] [wl res 55f233b8af90] destroyed (wl_resource_destroy not sent)`

Does this mean anything to you?

**Comment by OrderlyUnicode** [comment]
It has been pointed out to me in private conversations about this issue, and confirmed in my own testing, that resolution effects VRAM usage quite a bit more than I thought it would. This is starting to seem like less of a software issue on my end and more a shortsided hardware decision on the part of the company behind my particular machine. I will not close this issue just yet, since @TristenSeth 's comments point to a more severe issue that I may have simply managed to avoid. However, if you feel that their issue is best left for a different issue alltogether, I will close this one.

**Comment by ashbir** [comment]
I have this issue too, on the startup the Hyprland use ~200MiB of vram. But after hibernate the vram usage is a lot less become ~50MiB after I use it several minutes.

### Startup

![swappy-20231129-062405](https://github.com/hyprwm/Hyprland/assets/5442293/cf5162eb-5eaf-40ed-b738-5408aa1bf443)

### After Hibernate 

![swappy-20231129-064156](https://github.com/hyprwm/Hyprland/assets/5442293/c27bee2c-fb1f-4b40-b820-96f57d0cade2)

**Comment by tchofy** [comment]
Just woke up my laptop from a suspend with this, didn't check how much it was using before the suspend.
![image](https://github.com/hyprwm/Hyprland/assets/64620317/d1e8b1f2-68c5-49f4-9b12-db27bd1fb7b3)

**Comment by tchofy** [comment]
Now I'm unsure if my issue is different - Vram goes up by 10mb for every new notification from swaync, as long as one isn't currently being displayed on screen, but if I'm recording then vram stops increasing.

**Comment by izmyname** [comment]
Does it still happen? Can't confirm myself

**Comment by DADA30000** [comment]
@vaxerski on my friend laptop  hyprland consumes 1.1 gigs of vram on integrated graphics of ryzen 5 3500u, hyprland from git on NixOS

**Comment by DADA30000** [comment]
On stable hyprland from nixpkgs it's only 250MB on his laptop, I haven't tested this on my machine because I'm currently using stable hyprland

**Comment by vaxerski** [comment]
"stable" means not much, I know -git "consumes" more vram, but it's actually just lazy freeing by the driver and will be freed if it's needed

**Comment by DADA30000** [comment]
> "stable" means not much, I know -git "consumes" more vram, but it's actually just lazy freeing by the driver and will be freed if it's needed

oh, ok then, and by stable I meant latest version from nixpkgs unstable repo, it's 0.41.2

**Comment by SteavenGamerYT** [comment]
![image](https://github.com/user-attachments/assets/30f0a6f9-eade-459a-b5ad-0dfa30542c5f)
nvidia-smi says this for me

**Comment by romanstingler** [comment]
if you have Nvidia cards, update your `egl-wayland` package 
there have been a lot of changes between 1.13 and 1.15
https://github.com/NVIDIA/egl-wayland/compare/1.1.13...1.1.15

**Comment by Avimitin** [comment]
Is there a way to debug this issue? My hyprland boot up with ~200MB VRAM usage and after 8-9 hours it increase to to 3GB.

![image](https://github.com/user-attachments/assets/c40226d5-7d36-4c66-84bb-99b4e8670994)

**Comment by vaxerski** [comment]
that looks like a leak. Maybe it's that one nvidia bug? is this bisectable within hl? or easily reproducible

**Comment by Avimitin** [comment]
Its hard and painful to get the root cause :(. I will try my best to try all my daily behavior and see which one will cause memory leak.

**Comment by MCPO-Spartan-117** [comment]
Anyone happen to turn their monitors off and on(disconnect and reconnect)? Vram usages jumps up 150MBs everytime i turn my monitor off and on regardless if it is in or outside Hyprland.

I have a AMD card(6700xt) and a friend of mine also has this issue with a Nvidia card(3080) and it increases his by 300MBs, we both have a single monitor that turns completely off, he runs a CUDA test that allocates all available Vram and barely any gets freed.

He has a suspicion that the DRM buffer doesn't get freed when the monitor turns off.

**Comment by MCPO-Spartan-117** [comment]
I'm using git and i think he is too.
We have patched Hyprlands but this happens on 'vanilla' too. 
[hyprland.log](https://github.com/user-attachments/files/16736376/hyprland.log)

**Comment by gnusenpai** [comment]
A simple way to reproduce this is to unplug and replug your **only** monitor after a few seconds. Then, boom, vram leak.

**Comment by gnusenpai** [comment]
I don't mean to spam this issue, but I did some additional testing and found that this affects multi-monitor users as well as single.

You just unplug any monitor and plug it back in. It results in unreclaimable* VRAM being allocated, with the amount seemingly dependent on monitor resolution. Apparently, a lot of modern monitors fully disconnect when turned off, put into standby, or DPMS'd, which also triggers this.

*(Confirmed to be unreclaimable on at least NVIDIA: VRAM usage stacks each time up until ~96%, CUDA programs can allocate less and less memory, programs eventually fail to start, existing ones have their buffer get stuck, etc. AMD/Intel testing welcome.)

**Comment by vaxerski** [comment]
hm, is this reproducible on aq-git and hl-git?

**Comment by h3xaguard** [comment]
See #7728
I believe I have had a very similar issue, and if not it is likely @Avimitin 's issue.
Just clone the source code, goto src/render/Renderer.cpp
Edit the file, search for `static int  driverMajor = 0;` and change it to `int  driverMajor = 0;`.
I believe that this is the fix to the NVIDIA VRAM issue.

**Comment by PancakeTAS** [comment]
This VRAM issue is actually very annoying..
I start Hyprland and it immediately uses 600MiB of VRAM, which I am absolutely fine with considering I have an ungodly amount of high resolution monitors..
So then I open firefox, resize it like it was mentioned here before and my memory usage explodes to 1700-1800MiB and it doesn't go lower. It doesn't go higher either, which is nice, but still nearly 2GiB of VRAM usage is not acceptable for most people..
It *never* goes lower by the way, I can close all applications, reload Hyprland and do whatever I want but it stays up there.
The KMS change did absolutely nothing and I really doubt this is an NVIDIA bug..

**Comment by vaxerski** [comment]
can you bisect this?

**Comment by Edgars-Cirulis** [comment]
> This VRAM issue is actually very annoying.. I start Hyprland and it immediately uses 600MiB of VRAM, which I am absolutely fine with considering I have an ungodly amount of high resolution monitors.. So then I open firefox, resize it like it was mentioned here before and my memory usage explodes to 1700-1800MiB and it doesn't go lower. It doesn't go higher either, which is nice, but still nearly 2GiB of VRAM usage is not acceptable for most people.. It _never_ goes lower by the way, I can close all applications, reload Hyprland and do whatever I want but it stays up there. The KMS change did absolutely nothing and I really doubt this is an NVIDIA bug..

Nvidia open? If yes I've also got vram issue on nvidia open meanwhile it dosnt happen on closed source drivers.

**Comment by PancakeTAS** [comment]
> can you bisect this?

yes, [aquamarine commit](https://github.com/hyprwm/Hyprland/commit/016da234d0e852de3ef20eb2e89ac58d2a85f6e7). though what interesting is that resizing firefox does end up with 1700-1800MiB again, but as soon as I close firefox it is freed again. perhaps there is some "buffer cache" in aquamarine somewhere?

> Nvidia open? If yes I've also got vram issue on nvidia open meanwhile it dosnt happen on closed source drivers.

i'm on closed and the issue exists, i'll quickly check open but I doubt it makes a difference

**Comment by vaxerski** [comment]
No, aquamarine does not manage vram, neither does hl. That's done by the driver.

**Comment by MCPO-Spartan-117** [comment]
> This VRAM issue is actually very annoying.. I start Hyprland and it immediately uses 600MiB of VRAM, which I am absolutely fine with considering I have an ungodly amount of high resolution monitors.. So then I open firefox, resize it like it was mentioned here before and my memory usage explodes to 1700-1800MiB and it doesn't go lower. It doesn't go higher either, which is nice, but still nearly 2GiB of VRAM usage is not acceptable for most people.. It _never_ goes lower by the way, I can close all applications, reload Hyprland and do whatever I want but it stays up there. The KMS change did absolutely nothing and I really doubt this is an NVIDIA bug..

Unfortunately, this does seem to be a Nvidia bug, i've done exactly what @gnusenpai does with firefox including the exact applications with the same Hyprland and Aquamarine versions with a 6700XT and i couldn't replicate the memory leak, GPU memory jumps up 200-300MBs while resizing and drops back down immediately after.

**Comment by Avimitin** [comment]
Maybe related https://github.com/hyprwm/Hyprland/issues/7362#issuecomment-2303849833

**Comment by jerezoff** [comment]
🤭
![изображение](https://github.com/user-attachments/assets/847a65c3-c616-43ff-b1f7-ca186cbf6fc7)

**Comment by seemsindie** [comment]
![image](https://github.com/user-attachments/assets/dc9516b1-ade7-4e98-a805-63677e880a63)
2.7GB Fresh install (today - 0.44.1)

**Comment by vaxerski** [comment]
@seemsindie 

![image](https://github.com/user-attachments/assets/fdc78097-e279-4c75-bc04-85f339c8349f)

I'm tired of linking to nvidia. It's a leak in their drivers

**Comment by PancakeTAS** [comment]
Is the CPU memory leak also part of the NVIDIA drivers fault?
![image](https://github.com/user-attachments/assets/a948010a-b6e7-47f9-95df-b74220b753cf)
Unless of course this is normal, in which case ignore me
(res, shr, priv)

**Comment by seemsindie** [comment]
@vaxerski Yeah I understand. I reported it on NVIDIA forums (i don't expect anything will happen) https://forums.developer.nvidia.com/t/driver-leaking-memory-wayland-hyprland/309627

**Comment by Edgars-Cirulis** [comment]
@seemsindie. https://github.com/NVIDIA/egl-wayland/issues/126#issuecomment-2379945259 Here's fix temporarily until they release new update because that bug is fixed already on upcoming release.

**Comment by izmyname** [comment]
> @seemsindie. [NVIDIA/egl-wayland#126 (comment)](https://github.com/NVIDIA/egl-wayland/issues/126#issuecomment-2379945259) Here's fix temporarily until they release new update because that bug is fixed already on upcoming release.
> 
> Rename kwin_wayland to Hyprland

New `egl-wayland` version was released. Does it mean - the issue is solved and can be closed?

**Comment by vaxerski** [comment]
Hello there!

This issue has been closed, as we are moving from issues to discussions. More about this can be found [here](https://wiki.hyprland.org/Contributing-and-Debugging/Issue-Guidelines/) and in #9854.

Firstly, please make sure this issue is still relevant in latest Hyprland. If so, we ask you to open a [discussion](https://github.com/hyprwm/Hyprland/discussions) (please read the discussion guidelines first, in the pinned post)

Thank you for your understanding!

**Comment by paboum** [comment]
Oh, he cannot fix it because his bubu webpage was spammed and couldn't either use cloudflare or Anubis.
Related: https://github.com/vaxerski/checkpoint/issues/36

**Comment by SteavenGamerYT** [comment]
this is why i don't use hyprland always leaves tings broken

**Comment by edgars-cirulis** [comment]
> this is why i don't use hyprland always leaves tings broken

My man

**Comment by SteavenGamerYT** [comment]
> > this is why i don't use hyprland always leaves tings broken
> 
> My man

sway just works

**Comment by DADA30000** [comment]
> > > this is why i don't use hyprland always leaves tings broken
> > 
> > 
> > My man
> 
> sway just works

well, ofc, less features = less bugs

**Comment by starbaser** [comment]
> this is why i don't use hyprland always leaves tings broken

\>omarchy user

**Comment by Zebra2711** [comment]
> this is why i don't use hyprland always leaves tings broken

@SteavenGamerYT Then why does `libwlroots` have the same config to work around this issue in the nvidia profile????
```json
{ "pattern" : { "feature" : "procname", "matches" : "plasmashell" },            "profile" : "No VidMem Reuse" },
{ "pattern" : { "feature" : "procname", "matches" : "cosmic-comp" },            "profile" : "No VidMem Reuse" },
{ "pattern" : { "feature" : "procname", "matches" : "Hyprland" },               "profile" : "No VidMem Reuse" },
{ "pattern" : { "feature" : "procname", "matches" : "Xwayland" },               "profile" : "No VidMem Reuse" },
{ "pattern" : { "feature" : "dso",      "matches" : "libkwin" },                "profile" : "No VidMem Reuse" },
{ "pattern" : { "feature" : "dso",      "matches" : "libmutter" },              "profile" : "No VidMem Reuse" },
{ "pattern" : { "feature" : "dso",      "matches" : "libwlroots" },             "profile" : "No VidMem Reuse" },
{ "pattern" : { "feature" : "dso",      "matches" : "libweston" },              "profile" : "No VidMem Reuse" },
```
Isn't this an nvidia issue?

**Comment by SteavenGamerYT** [comment]
> > this is why i don't use hyprland always leaves tings broken
> 
> [@SteavenGamerYT](https://github.com/SteavenGamerYT) Then why does `libwlroots` have the same config to work around this issue in the nvidia profile????
> 
> { "pattern" : { "feature" : "procname", "matches" : "plasmashell" },            "profile" : "No VidMem Reuse" },
> { "pattern" : { "feature" : "procname", "matches" : "cosmic-comp" },            "profile" : "No VidMem Reuse" },
> { "pattern" : { "feature" : "procname", "matches" : "Hyprland" },               "profile" : "No VidMem Reuse" },
> { "pattern" : { "feature" : "procname", "matches" : "Xwayland" },               "profile" : "No VidMem Reuse" },
> { "pattern" : { "feature" : "dso",      "matches" : "libkwin" },                "profile" : "No VidMem Reuse" },
> { "pattern" : { "feature" : "dso",      "matches" : "libmutter" },              "profile" : "No VidMem Reuse" },
> { "pattern" : { "feature" : "dso",      "matches" : "libwlroots" },             "profile" : "No VidMem Reuse" },
> { "pattern" : { "feature" : "dso",      "matches" : "libweston" },              "profile" : "No VidMem Reuse" },
> 
> Isn't this an nvidia issue?

wtf do you mean by nvidia
i use AMD GPU I USE RX 7600 XT

**Comment by DADA30000** [comment]
> > > this is why i don't use hyprland always leaves tings broken
> > 
> > 
> > [@SteavenGamerYT](https://github.com/SteavenGamerYT) Then why does `libwlroots` have the same config to work around this issue in the nvidia profile????
> > { "pattern" : { "feature" : "procname", "matches" : "plasmashell" },            "profile" : "No VidMem Reuse" },
> > { "pattern" : { "feature" : "procname", "matches" : "cosmic-comp" },            "profile" : "No VidMem Reuse" },
> > { "pattern" : { "feature" : "procname", "matches" : "Hyprland" },               "profile" : "No VidMem Reuse" },
> > { "pattern" : { "feature" : "procname", "matches" : "Xwayland" },               "profile" : "No VidMem Reuse" },
> > { "pattern" : { "feature" : "dso",      "matches" : "libkwin" },                "profile" : "No VidMem Reuse" },
> > { "pattern" : { "feature" : "dso",      "matches" : "libmutter" },              "profile" : "No VidMem Reuse" },
> > { "pattern" : { "feature" : "dso",      "matches" : "libwlroots" },             "profile" : "No VidMem Reuse" },
> > { "pattern" : { "feature" : "dso",      "matches" : "libweston" },              "profile" : "No VidMem Reuse" },
> > Isn't this an nvidia issue?
> 
> wtf do you mean by nvidia i use AMD GPU I USE RX 7600 XT

this issue is about nvidia...

**Comment by SteavenGamerYT** [comment]
> > > > this is why i don't use hyprland always leaves tings broken
> > > 
> > > 
> > > [@SteavenGamerYT](https://github.com/SteavenGamerYT) Then why does `libwlroots` have the same config to work around this issue in the nvidia profile????
> > > { "pattern" : { "feature" : "procname", "matches" : "plasmashell" },            "profile" : "No VidMem Reuse" },
> > > { "pattern" : { "feature" : "procname", "matches" : "cosmic-comp" },            "profile" : "No VidMem Reuse" },
> > > { "pattern" : { "feature" : "procname", "matches" : "Hyprland" },               "profile" : "No VidMem Reuse" },
> > > { "pattern" : { "feature" : "procname", "matches" : "Xwayland" },               "profile" : "No VidMem Reuse" },
> > > { "pattern" : { "feature" : "dso",      "matches" : "libkwin" },                "profile" : "No VidMem Reuse" },
> > > { "pattern" : { "feature" : "dso",      "matches" : "libmutter" },              "profile" : "No VidMem Reuse" },
> > > { "pattern" : { "feature" : "dso",      "matches" : "libwlroots" },             "profile" : "No VidMem Reuse" },
> > > { "pattern" : { "feature" : "dso",      "matches" : "libweston" },              "profile" : "No VidMem Reuse" },
> > > Isn't this an nvidia issue?
> > 
> > 
> > wtf do you mean by nvidia i use AMD GPU I USE RX 7600 XT
> 
> this issue is about nvidia...

okay
